src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,17 +357,65 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
+class Hook
+	constructor: (opts) ->
+		@runFn = opts.RUN
+		@rollbackFn = opts.ROLLBACK
+		@rollbackArgs = []
+		@executed = false
+
+	setArgument: (arg) ->
+		if @rollbackArgs.length < @rollbackFn.length
+			@rollbackArgs.push(arg)
+		else
+			throw new Error('Too many arguments supplied to rollback function')
+
+	run: (args...) ->
+		Promise.try =>
+			@runFn.apply(null, args)
+		.tap =>
+			if @rollbackArgs.length != @rollbackFn.length
+				throw new Error('registerRollback not supplied enough arguments')
+			@executed = true
+		# If we detect that some rollbacks are registered but the rollback function is not saturated, we add the missing number of arguments as null.
+		# The ROLLBACK function should take care of checking and performing the correct number of rollback actions.
+		.tapCatch =>
+			if not _.isEmpty(@rollbackArgs)
+				missingArgs = @rollbackFn.length - @rollbackArgs.length
+				for i in _.range(missingArgs)
+					@rollbackArgs.push(null)
+				@rollback()
+
+	rollback: ->
+		Promise.try =>
+			@rollbackFn.apply(null, @rollbackArgs)
+
 runHooks = (hookName, args) ->
 	Object.defineProperty args, 'api',
 		get: _.once ->
 			return api[args.request.vocabulary].clone(passthrough: _.pick(args, 'req', 'tx'))
 	hooks = args.request?.hooks?[hookName] || []
 	Promise.map hooks, (hook) ->
-		hook(args)
+		args.registerRollback = (args...) ->
+			for arg in args
+				hook.setArgument(arg)
+		hook.run(args)
+
+# The execution order of rollback actions is unspecified
+undoHooks = (request) ->
+	Promise.map _.flatMap(request.hooks, _.identity), (hook) ->
```

Comment:
I think this should be `_.flatten(request.hooks)`

Reply:
`flatten` does not work here as `request.hooks` is an object like `{ PHASE: [hooks] }`. The current behaviour is  equivalent to `_.flatten(_.values(request.hooks))`

Reply:
I think `_.flatten(_.values(request.hooks))` makes sense then, since it makes the intent clearer